### PR TITLE
adding pfam samples sweeps and images for pfam and gfp experiments

### PR DESCRIPTION
**Pull request by amirshane**

### CODE REVIEW

File: contextual_lenses/pfam_family_classification/analyze_pfam_sweeps.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""Analyze and plot pfam train samples sweeps."""
+
+
+import pandas as pd
+
+import matplotlib.pyplot as plt
+
+
+df = pd.read_csv('samples_sweep.csv')
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/analyze_pfam_sweeps.py:9]: Put this and the calls to sweep_plot below inside a main function, and do

```
if __name__ == "__main__":
  ...
```

Good practice so that the code isn't executed if you, for example, import the file into ipython notebook then call the function (in which case you don't want this default case to run)

https://stackoverflow.com/questions/419163/what-does-if-name-main-do

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors.py [hunk 2]
```diff
@@ -64,6 +64,74 @@
...
+for i in range(1, 401):
+  family_name = 'PF%05d' % i
+  train_family_accessions.append(family_name)
+  test_family_accessions.append(family_name)
+
+results = pfam_nearest_neighbors_classification(encoder=optimizer.target, 
+											    train_family_accessions=train_family_accessions, 
+											    test_family_accessions=test_family_accessions)[0]
+
+print("Lens training: None, NN training: PF00001 - PF00400")
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors.py:130]: consider making a function which accepts a start & end pfam family, then computes the above, rather than duplicating code three times:

```
def train_and_test_pfam_families(start, end, n_train_samples):
  # Make train & test families
  ... 
  #  run pfam_nearest_neighbors_classification
  ...
  # Print results
```

Could also be used in next file

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+named_family_accessions = {}
+named_family_accessions[0] = ('1-100', ['PF%05d' % i for i in range(1, 101)])
+named_family_accessions[1] = ('101-200', ['PF%05d' % i for i in range(101, 201)])
+named_family_accessions[2] = ('1-200', ['PF%05d' % i for i in range(1, 201)])
+
+train_samples_sweep = [i for i in range(1, 6)] + [5*i for i in range(2, 6)] + [25*i for i in range(2, 5)] + [None]
+
+for train_samples in train_samples_sweep:
+    pretrain = 0
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:82]: this is a constant - define above the loop (here and for next `pretrain = 1` case).

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    for i in range(3):
+        families, family_accessions = named_family_accessions[i]
+        results = pfam_nearest_neighbors_classification(encoder=optimizer.target, 
+											            train_family_accessions=family_accessions, 
+											            test_family_accessions=family_accessions,
+                                                        train_samples=train_samples)[0]
+        accuracy = results['1-nn accuracy']
+        datum = [families, train_samples, pretrain, accuracy]
```
Comment (dmrd) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:90]: Prefer a named tuple or dictionary to keep metadata around about what each field is.

Pandas can directly convert a list of dictionaries to a dataframe: https://stackoverflow.com/questions/20638006/convert-list-of-dictionaries-to-a-pandas-dataframe

sweep_df = pd.dataframe(sweep_data)

File: contextual_lenses/load_transformer.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+'''Utils for loading FlaxLM or FlaxBERT models or parameters.'''
```
Comment (dmrd) [inline contextual_lenses/load_transformer.py:1]: style nit: use """ instead of ''' for quoted comments

File: contextual_lenses/train_utils.py [hunk 6]
```diff
@@ -130,8 +136,48 @@ def apply(self, x, encoder_fn, encoder_fn_kwargs, reduce_fn, reduce_fn_kwargs,
...
+  if predict_fn_params is not None:
+    for fn_name in fn_names:
+      if '_' + str(len(fn_names)-1) in fn_name:
+        predict_fn_name = fn_name
+    loaded_params[predict_fn_name] = predict_fn_params
+    
+  if encoder_fn_params is not None:
+    for fn_name in fn_names:
+        if encoder_fn_ind in fn_name:
+            encoder_fn_name = fn_name
```
Comment (dmrd) [inline contextual_lenses/train_utils.py:166]: Suggest checking in each of these for loops that encoder_fn_name (or predict_fn_name in previous one) is only set once.

Something like:
```
encoder_fn_name = None
for ...
  if encoder_fn_in in fn_name:
    if encoder_fn_name is not None:
      raise ValueError('Multiple instances of encoder detected. %s' % fn_name)
```

Good to have additional error checking to make sure it works as expected.

File: contextual_lenses/load_transformer.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
+"""Utils for loading FlaxLM or FlaxBERT models or parameters."""
+
+
+import os
+import tensorflow as tf
+import gin
+from protein_lm import data, models
+
+
+def load_transformer_model(ckpt_dir, model_cls, domain=None):
```
Comment (davidBelanger) [inline contextual_lenses/load_transformer.py:10]: this function should definitely have test coverage. The easiest way to do this would be to make a new small model, compute the output on some dummy inputs, save the model to disk, make a new model by loading from file, and check that the output on the dummy inputs is the same.

File: contextual_lenses/pfam_family_classification/analyze_pfam_sweeps.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+import pandas as pd
+
+import matplotlib.pyplot as plt
+
+
+def sweep_plot(df, families, title):
+
+  df = df[df['families']==families]
+  df_rand = df[df['lens_train']==0]
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/analyze_pfam_sweeps.py:12]: fyi you can do df[~df.lens_train] and df[df.lens_train]

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py [hunk 9]
```diff
@@ -0,0 +1,126 @@
+"""Nearest neighbors classification sampling sweep
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:1]: migrating this file would be a great first step when migrating to caliban. Rather than doing this for loop in a single process, have each be it's own process in the cluster.

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+}
+cnn_max_pool_model = create_representation_model(encoder_fn=encoder_fn,
+                                                 encoder_fn_kwargs=encoder_fn_kwargs,
+                                                 reduce_fn=reduce_fn,
+                                                 reduce_fn_kwargs=reduce_fn_kwargs,
+                                                 num_categories=26,
+                                                 output_features=num_families,
+                                                 embed=True)
+
+optimizer = create_optimizer(cnn_max_pool_model, learning_rate=0., weight_decay=0.)
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:69]: why is the learning rate zero?

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+                 }
+        print(datum)
+        sweep_data.append(datum)
+        
+
+# Restore optimizer from checkpoint. 
+if restore_dir is not None:
+  optimizer = checkpoints.restore_checkpoint(ckpt_dir=restore_dir, target=optimizer)
+
+train_samples_sweep = [i for i in range(1, 6)] + [5*i for i in range(2, 6)] + [25*i for i in range(2, 5)] + [None]
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:103]: I don't understand what the difference between this for loop and the one above is. Why run things twice?

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 # Parse command line arguments.
 from parser import parse_args
-tpu_name, save_dir, restore_dir, use_pmap = parse_args()
+tpu_name, save_dir, restore_dir, use_pmap, restore_transformer_dir, bidirectional = parse_args()
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py:6]: not necessary to change, but in the future it would be cleaner to have parse_args() return a collections.namedtuple. This way, when you add more possible command line args you don't have to ever place that parse_args is called. You also remove the number of global variables. 

lense_config = parse_args()
...
foo(lens_config.use_pmap)

File: contextual_lenses/pfam_family_classification/pfam_transformer_nearest_neighbors_sample_sweep.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
+"""Nearest neighbors classification sampling sweep
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_transformer_nearest_neighbors_sample_sweep.py:1]: there is a considerable amount of shared code with pfam_nearest_neighbors_sample_sweep. Can't you set up an experiment driver that takes a model and runs the big loop. Merging these would be good because then there'd only be one thing to migrate to be caliban-friendly.

File: contextual_lenses/pfam_family_classification/pfam_utils.py [hunk 14]
```diff
@@ -80,11 +81,29 @@ def create_pfam_df(family_accessions, test=False):
   pfam_df['mod_family_accession'] = pfam_df.family_accession.apply(lambda x: mod_family_accession(x))
   pfam_df = pfam_df[pfam_df.mod_family_accession.isin(family_accessions)]
   pfam_df['index'] = pfam_df.family_id.apply(lambda x: family_id_to_index[x])
-  pfam_df['one_hot_inds'] = pfam_df.sequence.apply(lambda x: residues_to_one_hot_inds(pad_seq(x)))
+
+  pfam_df['one_hot_inds'] = pfam_df.sequence.apply(lambda x: residues_to_one_hot_inds('M' + x[:511])) # x[:512] # pad_seq(x)
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_utils.py:85]: can you add a comment explaining the motivation for this?

File: contextual_lenses/pfam_family_classification/pfam_utils.py [hunk 15]
```diff
@@ -80,11 +81,29 @@ def create_pfam_df(family_accessions, test=False):
...
 
 
+def create_pfam_seq_batches(family_accessions, batch_size, test=False, samples=None, epochs=1,
+                            drop_remainder=False, buffer_size=None, seed=0, random_state=0):
+  """Creates iterable object of Pfam data batches."""
+
+  pfam_df = create_pfam_df(family_accessions, test=test)
+    
+  if samples is not None:
+    pfam_df = pfam_df.sample(frac=1, replace=False, random_state=random_state)
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_utils.py:97]: can you add a comment explaining that this is shuffling?

File: contextual_lenses/pfam_family_classification/pfam_utils.py [hunk 16]
```diff
@@ -80,11 +81,29 @@ def create_pfam_df(family_accessions, test=False):
...
+    pfam_df = pfam_df.groupby('mod_family_accession').head(samples).reset_index()
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_utils.py:98]: remind me, what's mod_family_accession vs. family_accession?

File: contextual_lenses/pfam_family_classification/pfam_utils.py [hunk 17]
```diff
@@ -154,7 +173,7 @@ def compute_embeddings(encoder, data_batches):
 
 
 def pfam_nearest_neighbors_classification(encoder, train_family_accessions, test_family_accessions, batch_size=512, 
-                                          n_neighbors=1, train_samples=None, test_samples=None, restricted=False):
```
Comment (davidBelanger) [inline contextual_lenses/pfam_family_classification/pfam_utils.py:157]: The variable name 'restricted' is vague. What does it mean?

File: contextual_lenses/tape_gfp/tape_gfp_utils.py [hunk 18]
```diff
@@ -31,27 +33,24 @@ def dataset_to_df(in_name):
 
 
 # Padding.
-def pad_seq(seq, pad_char='-'):
```
Comment (davidBelanger) [inline contextual_lenses/tape_gfp/tape_gfp_utils.py:34]: too much overlap with the pfam utils. Shouldn't you just have utils.py for this shared code? Otherwise, they may fall out of sync.

File: contextual_lenses/train_utils.py [hunk 19]
```diff
@@ -34,18 +39,46 @@ def create_data_iterator(df, input_col, output_col, batch_size, epochs=1, buffer
...
+  """Returns whether or not layer name is contained in path."""
 
+  return layer in path
 
-def create_optimizer(model, learning_rate, weight_decay):
-  """Instantiates Adam optimizer."""
 
-  optimizer_def = optim.Adam(learning_rate=learning_rate, weight_decay=weight_decay)
-  optimizer = optimizer_def.create(model)
+def create_optimizer(model, learning_rate, weight_decay, layers=None):
```
Comment (davidBelanger) [inline contextual_lenses/train_utils.py:62]: Lots of flax details here. make sure ddohan signs off on this entire file. Also need to have some loading/saving tests. 
@dmrd

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py (hunk 10 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:69]: no learning is being done, optimizer is just being used to compute embeddings, could set lr to anything

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py (hunk 11 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:103]: this loop is for the trained lens

File: contextual_lenses/pfam_family_classification/pfam_utils.py (hunk 16 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_utils.py:98]: mod_family_accession just turns PFabcd.efg into PFabcd (removes decimal and things after)

File: contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py (hunk 9 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_nearest_neighbors_sample_sweep.py:1]: agreed, working on that now

File: contextual_lenses/pfam_family_classification/pfam_utils.py (hunk 17 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_utils.py:157]: not sure why this is here, this 'restricted' variable has been removed

File: contextual_lenses/tape_gfp/tape_gfp_utils.py (hunk 18 above)
Comment (amirshane) [reply contextual_lenses/tape_gfp/tape_gfp_utils.py:34]: pad_seq should not be here, the new method calls from protein_lm.domains, will look into this

File: contextual_lenses/pfam_family_classification/pfam_utils.py (hunk 14 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_utils.py:85]: oh yes, the 'M' + x[:511] was for prepending an 'M', I've removed this for now but we may want to add it back later

File: contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py (hunk 12 above)
Comment (amirshane) [reply contextual_lenses/pfam_family_classification/pfam_one_hot_lenses.py:6]: great idea, the growing list of command-line arguments was worrying me as well
